tests/test_updater_ng.py

```diff
@@ -316,6 +318,25 @@ def test_missing_targetinfo(self) -> None:
         # Get targetinfo for non-existing file
         self.assertIsNone(self.updater.get_targetinfo("file33.txt"))
 
+    @patch.object(os, "replace", wraps=os.replace)
+    @patch.object(os, "remove", wraps=os.remove)
+    def test_persist_metadata_fails(
+        self, wrapped_remove: MagicMock, wrapped_replace: MagicMock,
+    ) -> None:
+        # Testing that when write succeeds (the file is created) and replace
+        # fails by throwing OSError, then the file will be deleted.
+        wrapped_replace.side_effect = OSError()
+        with self.assertRaises(OSError):
+            self.updater._persist_metadata("target", b"data")
+
+        wrapped_replace.assert_called_once()
+        wrapped_remove.assert_called_once()
+
+        # Assert that the created tempfile during writing is eventually deleted.
+        encoded_name = parse.quote("target", "")
+        filepath = os.path.join(self.updater._dir, f"{encoded_name}.json")
+        self.assertFalse(os.path.exists(filepath))
```

Comment:
are you sure this checks what the comment claims?

Reply:
Yes, because I am constructing the filepath the same way it's done in `updater._persist_metadata`.

Reply:
`filepath` seems to be the final filename, not the temporary filename. comment claims this checks that tempfile is deleted.

Why not actually list all the files in the directory and check that this matches your expectations?

Reply:
I see what you mean. Right now my code checks if the filename is removed, but we want to check if the temporary name exists. The temporary name has a prefix `tmp` and random letters.
I will list the directory content and check for a file with a `tmp` prefix.